### ISSUE TITLE
Python Translation Initial Results

### ISSUE THREAD

**Issue by andersonfrailey** [open-source-economics/taxdata#92]
I just finished running the python-produced PUF file through tax-calc. [Here](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/New%20PUF%20Analysis.ipynb) is a notebook containing the results. Is there anything else people would like added to the notebook?

Most of the results are almost exactly the same. There is a slight dip in the number of returns and tax units with zero/negative income and combined tax liability. The latter is connected to the former because if you look at the number of units with zero/negative tax liability as a percent of the total returns they are similar. The drop in total number tax-units is also almost entirely non-filers, the sum of which is cut roughly in half.

The top decile also sees a noticeable increase in tax liability. I haven't pinned down the root cause of this yet. If you look at the distribution of tax liability it's actually almost exactly the same across the board.

There's one change I need to make to the scripts that adds a variable containing the number of dependents under 18 to address [Tax-Calc issue #1409](https://github.com/open-source-economics/Tax-Calculator/issues/1409). After that, I'll open a PR to merge everything in when I'm finished making changes to `puf_data/finalprep.py` and `puf_stage2/stage2.py`.

@MattHJensen @martinholmer @Amy-Xu @hdoupe

**Comment by martinholmer** [comment]
@andersonfrailey, Please coordinate with me before you issue a new Python-generated version of the `puf.csv` file because I want to add some code to the `final_prep.py` script that hopefully will eliminate the need in Tax-Calculator for the out-of-place `ID_Casualty_frt` data adjustment.

@MattHJensen

**Comment by martinholmer** [comment]
@andersonfrailey,  Whey are there only **half** the number of non-filers in the Python-generated file relative to the old SAS-generated file?
```
filers_py = (calc_py.records.filer * calc_py.records.s006).sum()
filers_sas = (calc_sas.records.filer * calc_sas.records.s006).sum()
nonfilers_py = (calc_py.records.s006[calc_py.records.filer == 0]).sum()
nonfilers_sas = (calc_sas.records.s006[calc_sas.records.filer == 0]).sum()
print '# Filers, Python: {}'.format(filers_py)
print '# Filers, SAS:    {}'.format(filers_sas)
print '# Non-Filers, Python:  {}'.format(nonfilers_py)
print '# Non-Filers, SAS:    {}'.format(nonfilers_sas)
# Filers, Python: 153581159.34
# Filers, SAS:    153581143.76
# Non-Filers, Python:  8218694.52
# Non-Filers, SAS:    16838366.52
In [30]:
nonfilers_py / nonfilers_sas
Out[30]:
0.48809333792788684
```

**Comment by martinholmer** [comment]
@andersonfrailey,  Currently I'm making a study of Tax-Calculator utility functions that people use in their interactive work with the taxcalc package.  Your notebook in taxdata issue #92 is an impressive example of using the taxcalc package interactively.  Your script starts with these lines:
```
from taxcalc import *
from notebookfunctions import distribution, index_list
```
Can you explain the `distribution` and `index_list` functions you're using?  It seems as if you have rolled your own utility functions, which suggests an incompleteness in the offerings of utility functions in taxcalc.

**Comment by andersonfrailey** [comment]
@martinholmer, I'm happy to coordinate with you as @hdoupe and I put the finishing touches on the script. I'll be sure to keep you in the loop before any PUF releases.

I also share your concern with the large drop in the number of non-filers. Some variation is expected because filing status is randomly changed for some tax-units (see [this](https://github.com/open-source-economics/taxdata/blob/master/puf_data/StatMatch/3AdjustFILSTV2014.sas) file in the SAS scripts), but to see the total cut in half is an extreme result. I don't have an answer for this yet. I am looking through the scripts and intermediate files produced while it's running to see what I can find. I'll post all results as I find them in this thread.

The `distribution` function I wrote takes as arguments AGI, weight, and the variable you want to see the distribution of and returns two lists. One contains the weighted sum of the variable in each AGI bin, the other the percent of the total weighted sum of the specified variable in each bin. 

There are nineteen bins total because I wanted to match the breakdown found in table 1.4 of the [SOI Tax Stats](https://www.irs.gov/uac/soi-tax-stats-individual-statistical-tables-by-size-of-adjusted-gross-income).

The function `index_list` simply returns labels to be used as x-axis labels for the plots created from the distribution data.

I'm happy to send you/post to GitHub the code for both.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> The `distribution` function I wrote takes as arguments AGI, weight, and the variable you want to see the distribution of and returns two lists.  One contains the weighted sum of the variable in each AGI bin, the other the percent of the total weighted sum of the specified variable in each bin.

Thanks for the explanation.  Let me be sure I understand.
```
sas_was = distribution(calc_sas.records.e00200, calc_sas.records.s006, calc_sas.records.c00100)
```
So `sas_was` is a tuple containing two lists, right?

As part of my study of Tax-Calculator usage patterns, I'm curious to know if you used any functions in the `taxcalc/utils.py` file to build your `distribution` function.  Or, were those utility functions not helpful for some reason?  I guess those utility functions all pre-date the `agi_bin` variable recently added to the `puf.csv` file.  One final question: did you even look in `utils.py` to see if there was a helpful function, or was it just easier to roll your own special-purpose function?

**Comment by andersonfrailey** [comment]
@martinholmer 
> So `sas_was` is a tuple containing two lists, right?

Correct.

> I'm curious to know if you used any functions in the `taxcalc/utils.py` file to build your distribution function. Or, were those utility functions not helpful for some reason?

I did not use any of the utils functions, though in hindsight I could have taken advantage of the `weighted_sum` function.

>One final question: did you even look in `utils.py` to see if there was a helpful function, or was it just easier to roll your own special-purpose function?

In this case, it was easier to create my own function simply because I wanted to exactly match the SOI data format, rather than look by weighted decile or any of the other distribution options available in `utils.py`

**Comment by martinholmer** [comment]
> @andersonfrailey described his notebook.

This all makes perfect sense.  Thanks for answering my questions.

**Comment by martinholmer** [comment]
@andersonfrailey,  Can you show us the diagnostic tables for the python and sas results?  This will give us a better idea of any aggregate differences between the old sas-generated puf.csv and the new python-generated puf.csv input file.

So far, we know there is a big difference in the number of non-filers.  Do you have any sense about which is closer to the truth?

@MattHJensen @feenberg @Amy-Xu

**Comment by martinholmer** [comment]
@andersonfrailey, Can you show us the distribution of (taxpayer_age minus spouse_age) for the MARS==2 subsample?  One of the things we were looking for in the python-generated ages was a more sensible couple-age-difference distribution.  If the ages look better, then I could remove the "bandaid" logic in `puf_data/finalprep.py`.

@MattHJensen @feenberg @Amy-Xu

**Comment by andersonfrailey** [comment]
@martinholmer 
> So far, we know there is a big difference in the number of non-filers. Do you have any sense about which is closer to the truth?

I found that one of the wage thresholds used to determine filing status during the CPS tax-unit creation process was set at 1,000 rather than 10,000. Fixing this increased the number of non-filers by about 200,000. This still leaves a large deficit, but it's a start. I'm still combing through a few other sections of code related to filing status determination and will report back my findings.
 
> Can you show us the diagnostic tables for the python and sas results?

>Can you show us the distribution of (taxpayer_age minus spouse_age) for the MARS==2 subsample?

I will add both of these to the next iteration of the notebook.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> I found that one of the wage thresholds used to determine filing status during the CPS tax-unit creation process was set at 1,000 rather than 10,000. Fixing this increased the number of non-filers by about 200,000. This still leaves a large deficit, but it's a start. I'm still combing through a few other sections of code related to filing status determination and will report back my findings.
>
>> Can you show us the diagnostic tables for the python and sas results?
>> Can you show us the distribution of (taxpayer_age minus spouse_age) for the MARS==2 subsample?
>
> I will add both of these to the next iteration of the notebook.

Thanks.

**Comment by andersonfrailey** [comment]
[Notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/New%20PUF%20Analysis.ipynb) has been updated to include diagnostic tables and a histogram depicting the distribution of `age_head - age_spouse` as requested by @martinholmer.

The issue with the lack of non-filers has also been fixed.

When we make the tax-units and determine their filing status during the CPS tax unit creation process, we subject the units to a number of tests including income levels. In the SAS code, income is measured as the sum of wages, interest income, dividends, alimony, business income, pension, rent income, farm income, and unemployment compensation. We had defined income in the python file as all of that, plus social security thus increasing the income in many tax units to the point that they exceeded the threshold and were marked as filers. After fixing this, the number of non-filers shot up and is now roughly 97% of the total in the SAS file.

**Comment by martinholmer** [comment]
@andersonfrailey,

Great that you have the non-filer problem solved.  Good work!

Sorry I wasn't more clear about my age-difference request.  Please don't plot the absolute value of taxpayer_age minus spouse_age.  I'd like to see the distribution of both positive and negative values of the age difference.  Also, perhaps a table with the weighted counts to see if the new python-generated data really reduce some of the extreme values of age difference.  It looks like it does in the graph, but the counts are so small its hard to see in the graph.

**Comment by martinholmer** [comment]
@andersonfrailey,

Assuming the python-generated age-difference distribution looks good, how should we proceed?

In general it seems like you should prepare a pull request that processes this new raw python-generated `puf.csv` file.  This processing will probably produce slightly different stage2 weights and slightly different stage3 adjustment factors, right?  

And the `finalprep.py` logic will need to be changed to handle the better age variables.  In particular, assuming the age-difference distribution looks good, there is no need for the logic I added in taxdata pull request #87.  Is there other `finalprep.py` code that needs to be removed or altered as we process the python-generated raw `puf.csv` file?

Do you have any thoughts about [Tax-Calculator issue #1425](https://github.com/open-source-economics/Tax-Calculator/issues/1425), which should be dealt with as part of producing this new python-generated `puf.csv` file, if we do anything about that issue.

@MattHJensen @Amy-Xu @hdoupe

**Comment by andersonfrailey** [comment]
@martinholmer I've updated the [notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/New%20PUF%20Analysis.ipynb) to show the true value of the age differences rather than the absolute value. I also split the histograms and increased the number of bins to make them easier to read.

Are the tables I added to show the weighted counts in  various age difference bins along the lines of what you had in mind?

As you said, the next steps would include a PR that replaces the SAS scripts with Python in the `puf_data` directory and makes the needed changes to `stage2.py`  and `finalprep.py` (addressed below).

>This processing will probably produce slightly different stage2 weights and slightly different stage3 adjustment factors, right?

Correct. Most of the code will stay the same, but the tolerances will be slightly adjusted in `stage2.py` and the associated notebook. All other code in `stage2/3.py` will stay the same and only the output files will change.

In `finalprep.py`, a handful of changes would need to be made:
1. The `add_dependents` function is no longer needed. We adjusted the match code to count dependents in each category we track when creating tax units so those variables will be available immediately after the matching scripts run. This would make PR #93 unnecessary as well.
2. Assuming the age distribution is ok, we can remove the logic in PR #87. 
3. As can be seen in the notebook, there are a lot of variables not dropped during the match that should be added to `UNUSED_READ_VARS` in the `remove_unused_variables` function.
4. For now, `NEW_POST_PR83_UNUSED_READ_VARS` and the `remove_benefits_variables` functions are not needed. This initial iteration of the matching program does not add the additional benefits related variables that made those necessary. This will change when we do add them though, so simply commenting out the code that runs those functions would probably be best to avoid having to add it all back down the road.

As for [Tax-Calculator issue #1425](https://github.com/open-source-economics/Tax-Calculator/issues/1425), I believe your solution in Tax Data PR #94 is a good solution. 

cc @MattHJensen @Amy-Xu @hdoupe

**Comment by martinholmer** [comment]
@andersonfrailey said:
> I've updated the [notebook](http://nbviewer.jupyter.org/github/andersonfrailey/Notebook-Uploads/blob/master/New%20PUF%20Analysis.ipynb) to show the true value of the age differences rather than the absolute value. I also split the histograms and increased the number of bins to make them easier to read.

Thanks.  The graphs are cool looking and, even better, they show the new python-generated raw `cps_matched_puf.csv` file has quite sensible spouse age differences.  I'll begin working on a taxdata pull request that removes the *ad hoc* age-difference code added by pull request #87.

I'm in complete agreement about the changes in `puf_data/finalprep.py` that need to be made when we switch from using a SAS-generated to Python-generated `cps_matched_puf.csv` file.  But for now, @hdoupe and I need to concentrate on the narrower agenda in issue #96.

What's the answer to my question to you in #96?
```
What version of the raw SAS-generated cps-matched-puf.csv file are we currently using?
On my computer I have thiscps-matched-puf.csv file:

puf_data$ ls -l cps-m*
-rw-r--r--@ 1 mrh  staff  178345362 Mar 21 13:33 cps-matched-puf.csv

puf_data$ awk -F, 'NR>1{n++}END{print n}' cps-matched-puf.csv 
219806

It has eight fewer records than the current version of the puf.csv file.
Is that the file that the stages and finalprep work will begin with this week?
I want to make sure I have the same raw starting file as everybody else.
```

**Comment by andersonfrailey** [comment]
@martinholmer, the raw `cps-matched-puf.csv` file you're looking at (with 219,806 records) was created with the SAS scripts. We never established why running the scripts on my machine versus John's resulted in eight fewer records that the current puf.csv (there was a previous issue discussing this, but I haven't found it yet. I'll post a link when I do).

The raw file created from the python scripts are creating a PUF with 215,525 records and that's what we will be starting with.

I'll also post this in the other issue.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> The raw file created from the python scripts are creating a PUF with 215,525 records and that's what we will be starting with.

OK, that sounds good.  Can you send me a copy of that python-generated raw file via private channels?  I want to test my two pending pull requests (#94 and #97) with it.  Thanks.

**Comment by martinholmer** [comment]
@andersonfrailey said in taxdata issue #92:
> In finalprep.py, a handful of changes would need to be made:
>
> 1. The add_dependents function is no longer needed. We adjusted the match code to count dependents in each category we track when creating tax units so those variables will be available immediately after the matching scripts run. This would make PR #93 unnecessary as well.
>
> 2. Assuming the age distribution is ok, we can remove the logic in PR #87.
>
> 3. As can be seen in the notebook, there are a lot of variables not dropped during the match that should be added to UNUSED_READ_VARS in the remove_unused_variables function.
>
> 4. For now, NEW_POST_PR83_UNUSED_READ_VARS and the remove_benefits_variables functions are not needed. This initial iteration of the matching program does not add the additional benefits related variables that made those necessary. This will change when we do add them though, so simply commenting out the code that runs those functions would probably be best to avoid having to add it all back down the road.

@andersonfrailey, I've made all four of the above changes to the `puf_data/finalprep.py` script, but am wondering whether the current version of the `age_consistency` function is needed any longer.  Could you take a look and let me know what you think?

Other than this lingering issue, the Tax-Calculator repo is ready to go with the new Python-generated puf* files via [pull request 1429](https://github.com/open-source-economics/Tax-Calculator/pull/1429).

@MattHJensen

**Comment by andersonfrailey** [comment]
@martinholmer I believe we should keep the `age_consistency` function. I just reran the final prep script without it and the age distribution, while still better than the current SAS file, does have some extreme values. `age_head - `age_spouse` ranges from -58 to 85 rather than the -17 to 18 range achieved with the `age_consistency` function.

![agesas](https://user-images.githubusercontent.com/20684675/27134026-3cd59ac8-50e2-11e7-8929-bd8f0535a14f.jpeg)

![agepy](https://user-images.githubusercontent.com/20684675/27134041-5137072c-50e2-11e7-9e3e-1978a0f36abd.png)

**Comment by martinholmer** [comment]
@andersonfrailey,  I'm totally confused.  I ran the current (master branch) version of finalprep.py, which includes the age_consistency function **without** the PR#87 code, and get an age-difference range from -58 to +85.  What version of the age_consistency function did you use to generate an age-diff range of -17 to +18?  Did you use an age_consistency function that contained the #87 logic (which I removed yesterday in #97)?

**Comment by andersonfrailey** [comment]
@martinholmer that's my mistake. I was still on my branch with #87 logic included when I thought I was running without it. After correcting that mistake I get the same results as you do.

**Comment by martinholmer** [comment]
@andersonfrailey said:
> I was still on my branch with #87 logic included when I thought I was running without it. After correcting that mistake I get the same results as you do.

Thanks for the clarification; I had guessed that was the situation.

**Comment by martinholmer** [comment]
taxdata issue #92 is now resolved with the merge of taxdata pull request #100 (and associated changes in the `puf_data/finalprep.py` script).
